### ISSUE TITLE
Nested `transformed` calls on the root frame are allowed

### ISSUE THREAD

**Issue by Lisoph** [KevinKelley/nanovg-rs#71]
Consider the following code:

```Rust
extern crate nanovg as nvg;

use nvg::Transform;

fn main() {
    let ctx = nvg::ContextBuilder::new().build().unwrap();
    ctx.frame((10, 10), 1.0, |mut frame| {
        frame.transformed(Transform::new(), |mut f2| {
            frame.transformed(Transform::new(), |f3| {});
        });
    });
}
```

**What should the transformation for f3 be?**
- `frame * f2 * f3`: Very misleading, because `frame.transformed(Transform::new(), |f3| {});` would imply only `frame * f3`.
- `frame * f3`: Also misleading, because why is this possible to create a "sibling transform" as a "child transform"?

**Comment by KevinKelley** [comment]
I take it you're using a side-effect of the `mut` borrow, to prevent a second borrow, even though the outer frame is not actually going to be mutated?  

Not sure if I agree... `transformed` is, by design, not modifying the existing frame, but  creating a new sub-frame with a given transformation.  A widget with its own coordinate system, for example.  There's no reason a frame can't have many subframes... and many windows do indeed have multiple subwindows.

Maybe it's necessary to ensure that they're only worked on one-at-a-time.  But I wouldn't want to be required, unless it's truly not possible to, for example, handle subwindows each on an own thread or something...

The main confusion in the above example code, is from the second use of `frame` parameter, inside the innermost (where `f2` should normally have been used, we're a child of `f2` not of `frame`)...

**Comment by Lisoph** [comment]
I was trying to invalidate writing this:

```Rust
ctx.frame((10, 10), 1.0, |frame| {
    frame.transformed(Transform::new(), |f2| {});
    frame.transformed(Transform::new(), |f3| {});
});
```

as this:

```Rust
ctx.frame((10, 10), 1.0, |frame| {
    frame.transformed(Transform::new(), |f2| {
        frame.transformed(Transform::new(), |f3| {});
    });
});
```

Both snippets were valid and completely identitcal before my commit, but I viewed the second form as bad because it destroyed the semantic flow of transformations. Transformations are hierarchical and I want to reflect this property in code. The `transformed` method is perfect for this, as it creates a form of encapsulation (the closure) that visually implies a hierarchy (nesting). The second snippet escapes this philosophy.

In other words: All drawing operations inside a `transformed` closure **must** be transformed by the passed transformation. One is free to create sub-frames, that are transformed relative to the parent frame, since that is additive behaviour: Each `transformed` *adds* to the parent transform.

tl;dr: If you're inside a `transformed` closure, all you should see and know of the world is through the newly created `Frame`, passed to the closure. The way to implement this was by making `transformed` take &mut self, which disallows calling other methods on the parent transform inside a closure.

---

Sorry for the wall of text, I hope it made some sense.

**Comment by KevinKelley** [comment]
Not sure what you're seeing as a bad thing... your example is using Transform::new() meaning identity-transform, so not doing anything.  In the first form, f2 and f3 both inherit the same outer-transform; in the second form, f3 inherits the parent f2, but f2 is identity, so happens to be accidentally the same thing.

Are you worried about the lexical access, in the second form, where the code inside the f3 transform can still access the outer variables (`frame`)?  Still not seeing the problem, I guess; I mean yes it's true, but it's still code owned and written by someone who presumably knows what they want to be doing, right?  Not like there's any principles being violated here.

**Comment by KevinKelley** [comment]
By analogy: when you talk to your parents, you say things in a way they'll understand (f3 -> f2 transform).  If you're visiting your grandparents, you'll use somewhat different language (f3 -> f2 -> frame), but you can still talk to them.

**Comment by Lisoph** [comment]
> In the first form, f2 and f3 both inherit the same outer-transform; in the second form, f3 inherits the parent f2, but f2 is identity, so happens to be accidentally the same thing.

In both forms, f3 inherits from frame, so does f2. f3 does not inherit from f2 in the secon form, that's the problem.
<pre>
<b>frame</b>.transformed(Transform::new(), |<b>f2</b>| { // Parent = frame
    <b>frame</b>.transformed(Transform::new(), |<b>f3</b>| {}); // Parent = frame
});
</pre>
The code to inherit from f2 would look like: **f2**.transformed(Transform::new(), |**f3**| {});

That's exactly what this issue / my commit was all about: To remove this potential, nasty confusion by disallowing access to any parent frame (`frame.`).

**Comment by KevinKelley** [comment]
Yes, I understand that, I just don't think it's something that will happen often (how often do you write nested frames like that in the first place?); and, I can imagine situations where you might want the ability, so a hard "disallow" seems like too much, to prevent a possible typo-style error.

**Comment by Lisoph** [comment]
I try to follow a philosophy: Make interfaces easy to use correctly and hard to use incorrectly. It's something I adapted after following Scott Meyers for a bit. He has a great talk on this subject: [link](https://www.youtube.com/watch?v=5tg1ONG18H8).

> I can imagine situations where you might want the ability

Personally I can't. Could you give an example? If it's about being able to represent any transform at any point in time, that's still possible; you can draw anything anywhere with `transformed`, whilst still preserving the painters algorithm. Functionality wise, nothing changed (or should have, otherwise that's a bug).

**Comment by KevinKelley** [comment]
The example that popped into my head is one that, probably still won't work.  ...parceling out subwindow updates to separate threads, so that if one blocks the others can still update.... but still, everything has to funnel through the single display driver, and so therefore the "single UI thread" model probably still has to hold.  So, your change here is probably valid.

**Comment by Lisoph** [comment]
Yeah, multithreaded graphics is more or less impossible, since neither NanoVG or even OpenGL itself were designed with multithreading in mind.

**Comment by KevinKelley** [comment]
Yup... I was wondering,though, if, just maybe, the actual draw-calls to the GL hardware, could maybe be synchronized/blocking, inside a mutex or lock; and the user-level window-painting, inside our `frame` method, still be multithreaded.  Then a lot of the subwindow composition could possibly gain from multithreading, but still block on the actual dump to hardware of the GL display-list...

But that's just speculation, probably wishful thinking, and probably not really worth the trouble.  

Good call, no objections...
